File: backend/src/main/java/com/cvesters/aurevanta/tenant/Slug.java
package com.cvesters.aurevanta.tenant;

/**
 * What an organisation's handle may be, and how to build the next one along.
 *
 * <p>
 * Deriving a handle from a display name used to live here and now lives in the frontend,
 * because it stopped being a rule and became a <em>suggestion</em>: the handle is a field
 * somebody fills in, and proposing a value while they type is what a form does. What
 * stays here is the part the server is answerable for — the shape a handle must have, and
 * the alternative offered when the one they chose is taken.
 *
 * <p>
 * {@link #PATTERN} is the contract between the two. A proposal the server would refuse is
 * a bug in the proposer, not a disagreement about the rule.
 */
public final class Slug {

	/**
	 * Lower-case letters and digits in groups separated by single hyphens: no leading,
	 * trailing or doubled hyphen, and nothing that would need escaping in a URL. Held as
	 * a string because {@code @Pattern} needs one.
	 */
	public static final String PATTERN = "^[a-z0-9]+(-[a-z0-9]+)*$";

	/** Two, so a handle is still something a person could say out loud. */
	public static final int MIN_LENGTH = 2;

	/** Matches {@code tenants.slug}. */
	public static final int MAX_LENGTH = 80;

	private Slug() {
	}

	/**
	 * The {@code n}th alternative to a handle, shortened if the suffix would not
	 * otherwise fit the column.
	 */
	public static String withSuffix(String base, int n) {
		String suffix = "-" + n;
		int room = MAX_LENGTH - suffix.length();
		return ((base.length() <= room) ? base : base.substring(0, room)) + suffix;
	}

	/**
	 * The handle a suggestion should count from.
	 *
	 * <p>
	 * Somebody refused {@code acme-2} is offered {@code acme-3} rather than
	 * {@code acme-2-2}: they are plainly already counting, and a suggestion that appends
	 * to their count instead of continuing it reads as a mistake.
	 */
	public static String base(String handle) {
		// Never empty for a handle of the shape PATTERN allows: it has to start with a
		// letter or a digit, and only a trailing count is taken off. "42" keeps itself.
		return handle.replaceAll("-\\d+$", "");
	}

}

File: frontend/src/tenant/slug.ts
/** Length and shape the server enforces; mirrored so the browser stops what it would refuse. */
export const MINIMUM_SLUG_LENGTH = 2;
export const MAXIMUM_SLUG_LENGTH = 80;

/**
 * Proposes a handle for an organisation from the name somebody is typing.
 *
 * A *proposal*, which is why it lives here rather than on the server: the handle is a
 * field its owner fills in, and offering a value while they type is what a form does. The
 * server enforces the shape and knows nothing about names — so a proposal it would refuse
 * is a bug in here, not a disagreement about the rule.
 *
 * Folds accents, lowercases, and reduces anything that is not a letter or digit to a
 * single hyphen. A name with nothing usable in it proposes nothing, and the person types
 * their own: an empty required field asks a question, where a plausible default invites
 * somebody to accept it without reading it.
 */
export function proposeSlug(name: string): string {
  const folded = name
    .normalize('NFD')
    .replace(/\p{M}+/gu, '')
    .toLowerCase();
  const hyphenated = folded.replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '');
  return hyphenated.length <= MAXIMUM_SLUG_LENGTH
    ? hyphenated
    : hyphenated.slice(0, MAXIMUM_SLUG_LENGTH).replace(/-+$/, '');
}
